database.py

```diff
@@ -0,0 +1,57 @@
+import sqlite3
+import datetime
+
+db_file = "timelord.db"
+
+def createLogTable():
+    # Create time log table
+    con = sqlite3.connect(db_file)
+    cur = con.cursor()
+    cur.execute("CREATE TABLE IF NOT EXISTS time_log(user_id TEXT NOT NULL, selected_date date, minutes INTEGER NOT NULL)")
+    con.close()
+
+class SQLConnection:
+    def __init__(self):
+        # Open SQL connection
+        self.con = sqlite3.connect(db_file)
+        self.cur = self.con.cursor()
+
+    def __del__(self):
+        # Close SQL connection (saving changes to file)
+        self.con.close()
+
+    def insert_timelog_entry(self, user_id, selected_date, minutes):
+        # Dates are stored in plain text - SQLite doesn't have a specific date type. This still works and is sortable as long as
+        # dates are stored in the YYYY-MM-DD format (highest to lowest weight)
+
+        # SQLite3 documentation says this format should be used rather than formatted strings to prevent sql injection attacks
+        # This probably isn't necessary here but there's no good reason not to
+        self.cur.execute("INSERT INTO time_log VALUES (?,?,?);", (user_id, selected_date, minutes))
+        self.con.commit()
+
+    # Get the time logged by all users
+    def timelog_table(self):
+        res = self.cur.execute("SELECT * FROM time_log;")
+        return(res.fetchall())
+
+    # Get total minutes logged by user with given user_id
+    def time_sum(self, user_id):
+        # If the user has entries in the database return their total time logged, otherwise return 0
+        res = self.cur.execute(f"SELECT SUM(minutes) FROM time_log WHERE user_id = ?;", (user_id,))
+        minutes = res.fetchone()[0]
+        if (minutes != None):
+            return(minutes)
+        else:
+            return(0)
```

Comment:
Sorry mate I didn't explain this very well. Looks like `fetchone()` can return `None` which means that taking the 0'th index without checking is going to throw an exception. This is what I was after here, to guarantee a successful request before indexing the tuple.

```suggestion
        minutes = res.fetchone()
        if (minutes != None):
            return(minutes[0])
        else:
            return(0)
```

Reply:
Maybe I'm missing something but I don't think this works. If the user doesn't have entries in the log table, then `fetchone()` returns the tuple `(None,)`. If you make `minutes = res.fetchone()`, then `minutes != None` will return true because minutes will be a tuple object, not the None object itself. Then the None object at index 0 will be returned to the function in app.py and it will throw an exception when we check if it's greater than 0.

I've checked it against users with and without existing log entries and the way it is now works fine for both.

Reply:
The SQLITE3 module documentation does say `fetchone()` can return `None` but I think this is only for if the cursor has been moved forward beyond any existing rows (so for example any subsequent `fetchone()` calls on this particular cursor object will return None because there is only one row returned by `self.cur.execute(f"SELECT SUM(minutes) FROM time_log WHERE user_id = ?;", (user_id,))` although this row could just be a tuple containing None.

Reply:
Ahhh I see I see, my bad sorry!